### PR TITLE
Don't flatten builder data for single resource collection get request

### LINKED ISSUE #57459: `kubectl get -o json` swallows collection metadata & resourceVersion
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:
`kubectl get -o json nodes`
(not only nodes, also resourcequotas, probably any collection?)
outputs correct metadata for each entity, but empty metadata & resourceVersion, selfLink for the **collection**:
```
{
    "apiVersion": "v1",
    "items": [
        {
            "apiVersion": "v1",
            "kind": "Node",
            "metadata": {
                ...
                "resourceVersion": "20654",
                "selfLink": "/api/v1/nodes/localhost",
                "uid": "c5eeefc3-e4bf-11e7-be0a-5eba880b8555"
...
...
    ],
    "kind": "List",
    "metadata": {},
    "resourceVersion": "",
    "selfLink": ""
}
```

Same with `kubectl get -o yaml nodes`:
```
...
kind: List
metadata: {}
resourceVersion: ""
selfLink: ""
```

This is a client-side problem, because `kubectl get --loglevel=11 nodes` shows server does return collection metadata including resourceVersion:
```
I1220 16:15:06.517586   15855 request.go:991] Response Body: {"kind":"NodeList","apiVersion":"v1","metadata":{"selfLink":"/api/v1/nodes","resourceVersion":"20835"},"items":[...]}
```

**What you expected to happen**:
get -o json/yaml should display collection-level metadata.
Also note it tries to display `resourceVersion` & `selfLink` on top level, I'd expect them to be inside `metadata`, as received from server (and per API lists like https://v1-8.docs.kubernetes.io/docs/api-reference/v1.8/#nodelist-v1-core don't have `resourceVersion` & `selfLink` fields directly, they have `metadata`.

**How to reproduce it (as minimally and precisely as possible)**:
`kubectl get -o json nodes | tail`

**Anything else we need to know?**:

**Environment**:
Openshift (via minishift), but happens with `kubectl` and as shown above server responses are OK, it's kubectl client problem.

- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1+5115d708d7", GitCommit:"fff65cf", GitTreeState:"clean", BuildDate:"2017-08-09T21:39:40Z", GoVersion:"go1.9beta2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.1+5115d708d7", GitCommit:"fff65cf", GitTreeState:"clean", BuildDate:"2017-08-01T06:24:02Z", GoVersion:"go1.7.6", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: ThinkPad T450s laptop
- OS (e.g. from /etc/os-release):  Fedora 27.  
- Kernel (e.g. `uname -a`): `4.14.5-300.fc27.x86_64` on maching running kubectl, `4.4.41-boot2docker` inside minishift VM.
- Install tools:  kubectl from `origin-clients-3.6.0-1.fc27.x86_64` package, openshift 3.6.0 server from Minishift 1.9.0.

**Let me know if reproducing on real upstream kubernetes would help**

### PR DESCRIPTION
**Pull request by niuzhenguo**
**What this PR does / why we need it**:
For single resource collection get request, don't flatten the data from the builder. Leave the List as it is instead of aggregating the output into a virtual List object, which will swallows collection metadata & resourceVersion

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #57459 

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by smarterclayton** [comment]
If we have Flatten set (which we should) then len(infos) > 0 should be sufficient.  Did we lose flatten?

**Comment by niuzhenguo** [comment]
@smarterclayton With Flatten set, the original object is not passed to the visitor which means we already lost the collection metadata & resourceVersion

File: pkg/kubectl/resource/builder.go [hunk 1]
```diff
@@ -567,7 +567,9 @@ func splitResourceTypeName(s string) (resourceTuple, bool, error) {
 // compatible types into individual entries and give them their own items. The original object
 // is not passed to any visitors.
 func (b *Builder) Flatten() *Builder {
-	b.flatten = true
+	if len(b.resources) != 1 {
```
Comment (liggitt) [inline pkg/kubectl/resource/builder.go]: making this behave differently depending on the state of `Builder` isn't good

Comment (niuzhenguo) [reply pkg/kubectl/resource/builder.go]: @liggitt thanks! Aggree, should not change the Flatten() behavior. Will move the check to Get Run cmd func.

**Comment by niuzhenguo** [comment]
@liggitt Updated to call b.Flatten only if MultipleTypesRequested, please have a look again, thanks!

**Comment by dims** [comment]
/close

please reopen if necessary

**Pull request closed without merging**